services/catalog/src/simcore_service_catalog/api/routes/services.py

```diff
@@ -103,70 +122,56 @@ async def list_services(
-
-            # Write Access Granted: fill in the service rights
-            access_rights: List[
-                ServiceAccessRightsAtDB
-            ] = await services_repo.get_service_access_rights(
-                service.key, service.version, product_name=x_simcore_products_name
-            )
-            service.access_rights = {rights.gid: rights for rights in access_rights}
-
-            # Write Access Granted: override some of the values with what is in the db
-            service_in_db: Optional[
-                ServiceMetaDataAtDB
-            ] = await services_repo.get_service(service.key, service.version)
-            if not service_in_db:
-                logger.error(
-                    "The service %s:%s is not in the database",
-                    service.key,
-                    service.version,
-                )
-                continue
-
-            service = service.copy(
-                update=service_in_db.dict(exclude_unset=True, exclude={"owner"})
-            )
-            # the owner shall be converted to an email address
-            if service_in_db.owner:
-                service.owner = await groups_repository.get_user_email_from_gid(
-                    service_in_db.owner
-                )
+    # let's get all the services access rights
+    services_access_rights: Dict[
+        Tuple[str, str], List[ServiceAccessRightsAtDB]
+    ] = await services_repo.list_services_access_rights(
+        key_versions=list(services_in_db.keys()), product_name=x_simcore_products_name
+    )
 
-            detailed_services.append(service)
+    # let's get the service owners
+    services_owner_emails: Dict[
+        int, str
+    ] = await groups_repository.list_user_emails_from_gids(
+        [s.owner for s in services_in_db.values()]
+    )
 
-        # services = parse_obj_as(List[ServiceOut], data) this does not work since if one service has an issue it fails
-        except ValidationError as exc:
-            logger.warning(
-                "skip service %s:%s that has invalid fields\n%s",
-                detailed_metadata.get("key"),
-                detailed_metadata.get("version"),
-                exc,
-            )
+    # NOTE: for the details of the services:
+    # 1. we get all the services from the director-v0 (TODO: move the registry to the catalog)
+    # 2. we filter the services using the visible ones
+    # get the services from the registry and filter them out
+    filtered_services_in_registry = [
+        s
+        for s in list_frontend_services() + await director_client.get("/services")
+        if (s.get("key"), s.get("version")) in services_in_db
+    ]
 
-    return detailed_services
+    return [
+        _prepare_service_details(
+            s,
+            services_in_db[s["key"], s["version"]],
+            services_access_rights[s["key"], s["version"]],
+            services_owner_emails.get(services_in_db[s["key"], s["version"]].owner),
+        )
```

Comment:
Looks better. If I am correct you only make 3 queries to the DB in this function, right?

Reply:
no this one makes no query. everything is already available

Reply:
I was referring to the `list_services` function actually.

Reply:
then yes